<a href="https://colab.research.google.com/github/MUMADE-TADM/s1_python-rdvicen1/blob/main/Copia_de_TADM2022_S03b_Consultas_SQL_(Clase).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Consultas SQL


Antes de empezar cargamos nuestro archivo data_berka en nuestro entorno: [enlace](https://sorry.vse.cz/~berka/challenge/pkdd1999/berka.htm)

In [1]:
!gdown --id 1Osmr7qVN0NMuqYxvLesNn7H4nqPCVbsU

/usr/local/lib/python3.7/dist-packages/gdown/cli.py:131: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  category=FutureWarning,
Downloading...
From: https://drive.google.com/uc?id=1Osmr7qVN0NMuqYxvLesNn7H4nqPCVbsU
To: /content/data_berka.db
100% 70.2M/70.2M [00:00<00:00, 139MB/s]


En primer lugar cargamos la extensión que nos va permitir manejar comandos sql directamente en colab


In [2]:
%load_ext sql

Y conectamos con la base de datos

In [3]:
%sql sqlite:///data_berka.db

'Connected: @data_berka.db'

Comprobamos esquema

para que nos diga las tablas que han en la base de datos

In [4]:
%%sql
SELECT name,type
FROM sqlite_master
WHERE type = 'table'

 * sqlite:///data_berka.db
Done.


[('DISTRICT', 'table'),
 ('ACCOUNT', 'table'),
 ('TRANS', 'table'),
 ('CLIENT', 'table'),
 ('DISPOSITION', 'table'),
 ('CREDIT_CARD', 'table'),
 ('LOAN', 'table'),
 ('PERMANENT_ORDER', 'table')]

## Consultas sobre Préstamos

1. Obtener todos los atributos que tiene un banco sobre prestamos que tiene el banco

(extra limitar a 20 registros)

%%sql para que toda la celda sea sql
si solo un % seria solo una linea
; * para seleccionar toda la table; LIMIT lmitar

In [5]:
%%sql
SELECT *
FROM LOAN
LIMIT 20

 * sqlite:///data_berka.db
Done.


[(5314, 1787, 930705, 96396, 12, 8033, 'B'),
 (5316, 1801, 930711, 165960, 36, 4610, 'A'),
 (6863, 9188, 930728, 127080, 60, 2118, 'A'),
 (5325, 1843, 930803, 105804, 36, 2939, 'A'),
 (7240, 11013, 930906, 274740, 60, 4579, 'A'),
 (6687, 8261, 930913, 87840, 24, 3660, 'A'),
 (7284, 11265, 930915, 52788, 12, 4399, 'A'),
 (6111, 5428, 930924, 174744, 24, 7281, 'B'),
 (7235, 10973, 931013, 154416, 48, 3217, 'A'),
 (5997, 4894, 931104, 117024, 24, 4876, 'A'),
 (7121, 10364, 931110, 21924, 36, 609, 'A'),
 (6077, 5270, 931122, 79608, 24, 3317, 'A'),
 (6228, 6034, 931201, 464520, 60, 7742, 'B'),
 (6356, 6701, 931208, 95400, 36, 2650, 'A'),
 (5523, 2705, 931208, 93888, 36, 2608, 'A'),
 (6456, 7123, 931209, 47016, 12, 3918, 'A'),
 (7104, 10320, 931213, 259740, 60, 4329, 'A'),
 (6820, 9034, 931216, 38148, 12, 3179, 'A'),
 (6876, 9236, 931221, 86616, 12, 7218, 'A'),
 (7097, 10266, 931223, 75624, 24, 3151, 'B')]

2. Obtener los campos de fecha (date), amount (cantidad), duración (duration), pagos (payments), estado (status)

(extra limitar a 20 registros)

In [6]:
%%sql
SELECT date, amount, duration, payments, status
FROM LOAN
ORDER BY amount desc
LIMIT 20

 * sqlite:///data_berka.db
Done.


[(971019, 590820, 60, 9847, 'C'),
 (980123, 566640, 60, 9444, 'C'),
 (971112, 541200, 60, 9020, 'D'),
 (950217, 538500, 60, 8975, 'C'),
 (980120, 504000, 60, 8400, 'C'),
 (980522, 495180, 60, 8253, 'C'),
 (941219, 482940, 60, 8049, 'D'),
 (970212, 475680, 48, 9910, 'C'),
 (970907, 473280, 60, 7888, 'D'),
 (971225, 468060, 60, 7801, 'C'),
 (970613, 466608, 48, 9721, 'C'),
 (971227, 465504, 48, 9698, 'D'),
 (950709, 465072, 48, 9689, 'C'),
 (931201, 464520, 60, 7742, 'B'),
 (971105, 460980, 60, 7683, 'D'),
 (960723, 444864, 48, 9268, 'C'),
 (980829, 437460, 60, 7291, 'C'),
 (980213, 428784, 48, 8933, 'C'),
 (961118, 421008, 48, 8771, 'C'),
 (951029, 419880, 60, 6998, 'C')]

3. Obtener los distintos tipos de estado que se presentan en un préstamo.

distinct para que solo te saque datos distintos

In [7]:
%%sql
SELECT DISTINCT status
FROM LOAN

 * sqlite:///data_berka.db
Done.


[('B',), ('A',), ('C',), ('D',)]

4. Obtener solo los prestamos ya terminados.

para obtener en a y b

In [8]:
%%sql
SELECT *
FROM LOAN
WHERE status in ("A", "B")

 * sqlite:///data_berka.db
Done.


[(5314, 1787, 930705, 96396, 12, 8033, 'B'),
 (5316, 1801, 930711, 165960, 36, 4610, 'A'),
 (6863, 9188, 930728, 127080, 60, 2118, 'A'),
 (5325, 1843, 930803, 105804, 36, 2939, 'A'),
 (7240, 11013, 930906, 274740, 60, 4579, 'A'),
 (6687, 8261, 930913, 87840, 24, 3660, 'A'),
 (7284, 11265, 930915, 52788, 12, 4399, 'A'),
 (6111, 5428, 930924, 174744, 24, 7281, 'B'),
 (7235, 10973, 931013, 154416, 48, 3217, 'A'),
 (5997, 4894, 931104, 117024, 24, 4876, 'A'),
 (7121, 10364, 931110, 21924, 36, 609, 'A'),
 (6077, 5270, 931122, 79608, 24, 3317, 'A'),
 (6228, 6034, 931201, 464520, 60, 7742, 'B'),
 (6356, 6701, 931208, 95400, 36, 2650, 'A'),
 (5523, 2705, 931208, 93888, 36, 2608, 'A'),
 (6456, 7123, 931209, 47016, 12, 3918, 'A'),
 (7104, 10320, 931213, 259740, 60, 4329, 'A'),
 (6820, 9034, 931216, 38148, 12, 3179, 'A'),
 (6876, 9236, 931221, 86616, 12, 7218, 'A'),
 (7097, 10266, 931223, 75624, 24, 3151, 'B'),
 (4959, 2, 940105, 80952, 24, 3373, 'A'),
 (6499, 7401, 940105, 80952, 24, 3373, 'A'),
 (5479, 2486, 940110, 24516, 12, 2043, 'A'),
 (6721, 8475, 940121, 21072, 24, 878, 'A'),
 (7259, 11111, 940131, 108144, 36, 3004, 'A'),
 (5285, 1603, 940206, 78936, 12, 6578, 'A'),
 (5189, 1166, 940207, 149040, 48, 3105, 'A'),
 (6650, 8073, 940217, 49320, 12, 4110, 'B'),
 (7226, 10940, 940223, 197748, 36, 5493, 'A'),
 (7262, 11135, 940301, 182628, 36, 5073, 'A'),
 (5519, 2694, 940308, 91632, 12, 7636, 'A'),
 (6577, 7753, 940311, 51696, 24, 2154, 'A'),
 (6509, 7445, 940315, 55368, 24, 2307, 'A'),
 (6520, 7485, 940316, 91200, 24, 3800, 'A'),
 (6699, 8330, 940330, 7656, 24, 319, 'A'),
 (6336, 6555, 940331, 212400, 36, 5900, 'A'),
 (6202, 5891, 940418, 65184, 12, 5432, 'A'),
 (7066, 10131, 940502, 215388, 36, 5983, 'B'),
 (6072, 5256, 940510, 193968, 36, 5388, 'A'),
 (5130, 813, 940511, 24312, 12, 2026, 'A'),
 (6678, 8173, 940520, 80340, 12, 6695, 'A'),
 (6998, 9833, 940521, 97632, 12, 8136, 'A'),
 (6580, 7769, 940522, 196080, 48, 4085, 'A'),
 (6316, 6473, 940526, 76908, 12, 6409, 'B'),
 (6512, 7454, 940526, 63312, 48, 1319, 'A'),
 (6407, 6930, 940526, 25944, 24, 1081, 'A'),
 (5002, 173, 940531, 104808, 12, 8734, 'A'),
 (6647, 8051, 940601, 208320, 48, 4340, 'A'),
 (7137, 10439, 940603, 151560, 24, 6315, 'A'),
 (6903, 9337, 940604, 52128, 24, 2172, 'A'),
 (5918, 4576, 940604, 300204, 36, 8339, 'A'),
 (6642, 8027, 940605, 77640, 12, 6470, 'A'),
 (6210, 5927, 940606, 87216, 48, 1817, 'B'),
 (6279, 6273, 940607, 87360, 24, 3640, 'A'),
 (7249, 11065, 940612, 69360, 48, 1445, 'A'),
 (6112, 5442, 940617, 174744, 24, 7281, 'B'),
 (6596, 7844, 940625, 91584, 24, 3816, 'A'),
 (5657, 3354, 940705, 4980, 12, 415, 'A'),
 (5533, 2762, 940705, 111384, 24, 4641, 'A'),
 (6232, 6051, 940708, 276660, 36, 7685, 'B'),
 (6888, 9265, 940711, 39576, 12, 3298, 'A'),
 (6066, 5228, 940711, 173808, 24, 7242, 'A'),
 (6004, 4937, 940723, 143904, 24, 5996, 'A'),
 (5126, 790, 940724, 208128, 48, 4336, 'B'),
 (7220, 10911, 940728, 131292, 36, 3647, 'A'),
 (6668, 8129, 940730, 74736, 36, 2076, 'A'),
 (5395, 2176, 940730, 299088, 36, 8308, 'B'),
 (6064, 5215, 940804, 14628, 12, 1219, 'A'),
 (6919, 9403, 940805, 120888, 36, 3358, 'A'),
 (6841, 9104, 940807, 172080, 48, 3585, 'A'),
 (6038, 5120, 940810, 323472, 48, 6739, 'A'),
 (6729, 8519, 940811, 74688, 12, 6224, 'A'),
 (7198, 10788, 940823, 45768, 12, 3814, 'A'),
 (6686, 8260, 940825, 224604, 36, 6239, 'A'),
 (5836, 4259, 940827, 79632, 24, 3318, 'A'),
 (7061, 10105, 940827, 121896, 24, 5079, 'A'),
 (5595, 3050, 940902, 82896, 12, 6908, 'A'),
 (5723, 3674, 940910, 23628, 12, 1969, 'A'),
 (6120, 5481, 940911, 209184, 48, 4358, 'A'),
 (6895, 9307, 940919, 41904, 12, 3492, 'A'),
 (5882, 4427, 940926, 87816, 12, 7318, 'A'),
 (5754, 3834, 940928, 23052, 12, 1921, 'A'),
 (6594, 7824, 940928, 189696, 24, 7904, 'B'),
 (6738, 8566, 940929, 230220, 36, 6395, 'B'),
 (6624, 7965, 940930, 71460, 12, 5955, 'A'),
 (6216, 5952, 940930, 141240, 24, 5885, 'A'),
 (6229, 6040, 941002, 11736, 24, 489, 'A'),
 (5887, 4448, 941003, 192744, 36, 5354, 'A'),

5. Ordenar lo anterior por duración y cantidad

extra: limitar a 20 elementos

In [9]:
%%sql
SELECT *
FROM LOAN
WHERE status in ("A", "B")
ORDER BY duration, amount DESC
LIMIT 100

 * sqlite:///data_berka.db
Done.


[(6934, 9494, 970209, 116832, 12, 9736, 'A'),
 (5002, 173, 940531, 104808, 12, 8734, 'A'),
 (4986, 97, 970810, 102876, 12, 8573, 'A'),
 (5375, 2106, 961003, 101520, 12, 8460, 'A'),
 (7253, 11079, 970426, 98304, 12, 8192, 'A'),
 (6998, 9833, 940521, 97632, 12, 8136, 'A'),
 (6612, 7907, 970831, 97392, 12, 8116, 'A'),
 (5314, 1787, 930705, 96396, 12, 8033, 'B'),
 (6995, 9814, 960926, 96168, 12, 8014, 'A'),
 (5464, 2419, 950130, 95808, 12, 7984, 'A'),
 (5643, 3273, 970921, 94512, 12, 7876, 'B'),
 (5377, 2116, 970804, 92400, 12, 7700, 'A'),
 (5519, 2694, 940308, 91632, 12, 7636, 'A'),
 (7046, 10049, 970328, 91632, 12, 7636, 'A'),
 (5868, 4361, 970329, 90156, 12, 7513, 'A'),
 (7277, 11231, 971024, 89280, 12, 7440, 'A'),
 (4996, 132, 961106, 88440, 12, 7370, 'A'),
 (5882, 4427, 940926, 87816, 12, 7318, 'A'),
 (7119, 10361, 970602, 86880, 12, 7240, 'A'),
 (6876, 9236, 931221, 86616, 12, 7218, 'A'),
 (6921, 9422, 961029, 86184, 12, 7182, 'A'),
 (5366, 2062, 961029, 86184, 12, 7182, 'A'),
 (6150, 5622, 941102, 83016, 12, 6918, 'A'),
 (5595, 3050, 940902, 82896, 12, 6908, 'A'),
 (6678, 8173, 940520, 80340, 12, 6695, 'A'),
 (6059, 5196, 971228, 79824, 12, 6652, 'A'),
 (5285, 1603, 940206, 78936, 12, 6578, 'A'),
 (7051, 10068, 970728, 77808, 12, 6484, 'A'),
 (6642, 8027, 940605, 77640, 12, 6470, 'A'),
 (5117, 718, 970820, 76944, 12, 6412, 'A'),
 (6316, 6473, 940526, 76908, 12, 6409, 'B'),
 (5739, 3747, 950102, 76380, 12, 6365, 'A'),
 (5992, 4863, 970318, 76080, 12, 6340, 'A'),
 (6729, 8519, 940811, 74688, 12, 6224, 'A'),
 (5878, 4415, 970401, 74040, 12, 6170, 'A'),
 (6599, 7859, 961119, 72120, 12, 6010, 'A'),
 (6624, 7965, 940930, 71460, 12, 5955, 'A'),
 (5352, 1993, 950415, 68400, 12, 5700, 'B'),
 (5638, 3249, 970827, 68124, 12, 5677, 'A'),
 (6598, 7857, 950118, 68112, 12, 5676, 'A'),
 (6753, 8688, 970320, 67980, 12, 5665, 'A'),
 (5682, 3455, 970307, 67464, 12, 5622, 'A'),
 (7201, 10807, 970307, 67464, 12, 5622, 'A'),
 (6202, 5891, 940418, 65184, 12, 5432, 'A'),
 (5750, 3804, 960208, 63252, 12, 5271, 'A'),
 (5388, 2153, 961128, 61560, 12, 5130, 'A'),
 (5338, 1888, 970830, 57360, 12, 4780, 'B'),
 (7305, 11359, 960806, 54024, 12, 4502, 'A'),
 (5398, 2187, 971104, 53472, 12, 4456, 'A'),
 (5134, 825, 970202, 53076, 12, 4423, 'A'),
 (7284, 11265, 930915, 52788, 12, 4399, 'A'),
 (5856, 4320, 970809, 51360, 12, 4280, 'B'),
 (6650, 8073, 940217, 49320, 12, 4110, 'B'),
 (6456, 7123, 931209, 47016, 12, 3918, 'A'),
 (7198, 10788, 940823, 45768, 12, 3814, 'A'),
 (7199, 10789, 971209, 45768, 12, 3814, 'A'),
 (7264, 11141, 971007, 44940, 12, 3745, 'A'),
 (6137, 5574, 960410, 44628, 12, 3719, 'A'),
 (6144, 5593, 960809, 44376, 12, 3698, 'A'),
 (6315, 6472, 950206, 43164, 12, 3597, 'A'),
 (6621, 7957, 970614, 42900, 12, 3575, 'A'),
 (6727, 8505, 971210, 42840, 12, 3570, 'A'),
 (6535, 7546, 951102, 41952, 12, 3496, 'A'),
 (6895, 9307, 940919, 41904, 12, 3492, 'A'),
 (6888, 9265, 940711, 39576, 12, 3298, 'A'),
 (6887, 9263, 970508, 39576, 12, 3298, 'A'),
 (5634, 3229, 970501, 39024, 12, 3252, 'A'),
 (5358, 2018, 970121, 38520, 12, 3210, 'A'),
 (6820, 9034, 931216, 38148, 12, 3179, 'A'),
 (6362, 6738, 951031, 37908, 12, 3159, 'A'),
 (6306, 6425, 941220, 36684, 12, 3057, 'A'),
 (6929, 9472, 950508, 36396, 12, 3033, 'A'),
 (7008, 9897, 950922, 36168, 12, 3014, 'A'),
 (6310, 6440, 970617, 34572, 12, 2881, 'A'),
 (5932, 4626, 970618, 33348, 12, 2779, 'A'),
 (6758, 8742, 970729, 33144, 12, 2762, 'A'),
 (5576, 2962, 941221, 31248, 12, 2604, 'A'),
 (6501, 7418, 960429, 30276, 12, 2523, 'B'),
 (4961, 19, 960429, 30276, 12, 2523, 'B'),
 (6502, 7424, 971208, 30276, 12, 2523, 'A'),
 (4962, 25, 971208, 30276, 12, 2523, 'A'),
 (6865, 9199, 960911, 29448, 12, 2454, 'B'),
 (6578, 7765, 951211, 27984, 12, 2332, 'A'),
 (5005, 176, 970410, 27456, 12, 2288, 'A'),
 (5479, 2486, 940110, 24516, 12, 2043, 'A'),
 (5130, 813, 940511, 24312, 12, 2026, 'A'),
 (5723, 3674, 940910, 23628, 12, 1969, 'A'),
 (5229, 1329, 970906, 23520, 12, 1960, 'A'),
 (6878, 9242, 970124, 23184, 12, 1932

Cargar resultado de la sql en un DATAFRAME de Pandas

Ex. 1

>Plantear otro filtro para los préstamos (por cantidad, duración, status, etc.)


In [10]:
%%sql
SELECT date, amount, duration, payments, status
FROM LOAN
ORDER BY date DESC
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(981208, 240900, 60, 4015, 'C'),
 (981206, 139488, 24, 5812, 'C'),
 (981206, 55632, 24, 2318, 'C'),
 (981205, 352704, 48, 7348, 'C'),
 (981205, 52512, 12, 4376, 'C'),
 (981202, 230220, 36, 6395, 'C'),
 (981202, 160920, 36, 4470, 'C'),
 (981201, 163332, 36, 4537, 'C'),
 (981127, 276084, 36, 7669, 'C'),
 (981125, 91248, 48, 1901, 'C')]

 Ex. 2 

> Seleccionar fecha, cantidad, duración y pagos de aquellos préstamos que han sido terminados con una duracion mayor de 12 meses ordenados de mayor a menor cantidad de dinero

>extra: volcar en un DataFrame

In [11]:
%%sql
SELECT date, amount, duration, payments
FROM LOAN
WHERE status in ("A", "B")  AND duration >12
ORDER BY amount DESC
LIMIT 20

 * sqlite:///data_berka.db
Done.


[(931201, 464520, 60, 7742),
 (940810, 323472, 48, 6739),
 (940604, 300204, 36, 8339),
 (940730, 299088, 36, 8308),
 (950103, 288468, 36, 8013),
 (940708, 276660, 36, 7685),
 (930906, 274740, 60, 4579),
 (950108, 270648, 36, 7518),
 (950623, 265320, 36, 7370),
 (951028, 260028, 36, 7223),
 (931213, 259740, 60, 4329),
 (950428, 248148, 36, 6893),
 (951020, 231696, 36, 6436),
 (940929, 230220, 36, 6395),
 (941012, 227880, 36, 6330),
 (940825, 224604, 36, 6239),
 (951204, 217152, 36, 6032),
 (940502, 215388, 36, 5983),
 (940331, 212400, 36, 5900),
 (940911, 209184, 48, 4358)]

Cuantos registros cumplen esa condicion

In [12]:
%%sql
SELECT count(*)
FROM LOAN
WHERE status in ("A", "B")  AND duration >12

 * sqlite:///data_berka.db
Done.


[(131,)]

Prestamos no terminados en ejecucion con duracion entre 24 y 48 meses

In [13]:
%%sql
SELECT count(*)
FROM LOAN
WHERE status in ("C", "D")  AND duration between 24 and 48

 * sqlite:///data_berka.db
Done.


[(279,)]

## Group By

7. ¿Cuántos préstamos hay por cada estado y cual es la media de la cantidad de dinero prestado?

al lado de count se ponen el nombre, ponerle nombre para ordenar


tambien se puede poner con as

al ser una condicion sobre variables que se obtiene en la propia consulta hay que utilizar el having

In [14]:
%%sql
SELECT status, count(*) total, avg(amount)
FROM LOAN
WHERE duration >= 24
GROUP BY status
HAVING total >25
order by total desc

 * sqlite:///data_berka.db
Done.


[('C', 376, 180343.1489361702),
 ('A', 110, 122425.2),
 ('D', 44, 254127.27272727274)]

8. Por cada Región, quiero contar el número de ciudades y obtener el total del número de habitantes de la región



In [15]:
%%sql
SELECT A3, count(A2) no_districts, sum(A9) no_city, sum(A4) no_habi
FROM district
GROUP BY A3
HAVING no_city > 1

 * sqlite:///data_berka.db
Done.


[('central Bohemia', 12, 66, 1105234),
 ('east Bohemia', 11, 86, 1234781),
 ('north Bohemia', 10, 70, 1178977),
 ('north Moravia', 11, 60, 1970302),
 ('south Bohemia', 8, 48, 700595),
 ('south Moravia', 14, 82, 2054989),
 ('west Bohemia', 10, 69, 859306)]

otra forma de filtrar praga

In [16]:
%%sql
SELECT A3, count(A2) no_districts, sum(A9) no_city, sum(A4) no_habi
FROM district
WHERE NOT A3 = "Prague"
GROUP BY A3

 * sqlite:///data_berka.db
Done.


[('central Bohemia', 12, 66, 1105234),
 ('east Bohemia', 11, 86, 1234781),
 ('north Bohemia', 10, 70, 1178977),
 ('north Moravia', 11, 60, 1970302),
 ('south Bohemia', 8, 48, 700595),
 ('south Moravia', 14, 82, 2054989),
 ('west Bohemia', 10, 69, 859306)]

9. Contar ordenes permanentes (domiciliaciones) por banco y tipo de transacción

In [17]:
%%sql
SELECT bank_to, k_symbol, count(*) as c
FROM PERMANENT_ORDER
WHERE amount < 500
GROUP BY bank_to, k_symbol
ORDER BY bank_to
LIMIT 10

 * sqlite:///data_berka.db
Done.


[('AB', ' ', 20),
 ('AB', 'POJISTNE', 16),
 ('AB', 'SIPO', 17),
 ('CD', ' ', 34),
 ('CD', 'POJISTNE', 11),
 ('CD', 'SIPO', 18),
 ('EF', ' ', 23),
 ('EF', 'POJISTNE', 21),
 ('EF', 'SIPO', 10),
 ('EF', 'UVER', 1)]

# JOIN

10. ¿Cuántas cuentas de frecuencia mensual tienen un un préstamo activo?



para poder obtener cruce de datos de varias tablas hay que utilizar inner join

loan 1 para llamar asi la tabla, account a para llamar asi la tabla

In [18]:
%%sql
SELECT loan_id, "1".date loan_date, amount, duration, payments, frequency, a.date acc_date
FROM LOAN "1", ACCOUNT a
WHERE "1".account_id = a.account_id
AND status in ("A","B")
ORDER BY amount, a.date DESC
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(5657, 940705, 4980, 12, 415, 'monthly', 931105),
 (6699, 940330, 7656, 24, 319, 'monthly', 930328),
 (6019, 960725, 11400, 12, 950, 'monthly', 950803),
 (6229, 941002, 11736, 24, 489, 'monthly', 930422),
 (5103, 971128, 12540, 12, 1045, 'monthly', 960315),
 (5718, 970901, 14028, 12, 1169, 'monthly', 951028),
 (6064, 940804, 14628, 12, 1219, 'monthly', 930701),
 (5897, 960925, 15192, 24, 633, 'monthly', 941110),
 (5599, 950321, 15420, 12, 1285, 'monthly', 930523),
 (6377, 970317, 17184, 12, 1432, 'monthly', 960511)]

añadir datos demograficos:

Se añade district en from, campos qye quieres en select y en where and claves ajenas a.district = d.A1

In [19]:
%%sql
SELECT loan_id, A4, A9, A10, A11, "1".date loan_date, amount, duration, payments, frequency, a.date acc_date
FROM LOAN "1", ACCOUNT a, DISTRICT d
WHERE "1".account_id = a.account_id
AND a.district_id = d.A1
AND status in ("A","B")
ORDER BY amount, a.date DESC
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(5657, 86513, 5, 50.5, 8288, 940705, 4980, 12, 415, 'monthly', 931105),
 (6699, 78955, 9, 59.0, 9060, 940330, 7656, 24, 319, 'monthly', 930328),
 (6019, 387570, 1, 100.0, 9897, 960725, 11400, 12, 950, 'monthly', 950803),
 (6229, 1204953, 1, 100.0, 12541, 941002, 11736, 24, 489, 'monthly', 930422),
 (5103, 67298, 7, 43.8, 8561, 971128, 12540, 12, 1045, 'monthly', 960315),
 (5718, 107911, 7, 50.9, 8240, 970901, 14028, 12, 1169, 'monthly', 951028),
 (6064, 157042, 9, 33.9, 8743, 940804, 14628, 12, 1219, 'monthly', 930701),
 (5897, 122603, 8, 80.0, 8991, 960925, 15192, 24, 633, 'monthly', 941110),
 (5599, 128118, 8, 85.3, 9317, 950321, 15420, 12, 1285, 'monthly', 930523),
 (6377, 1204953, 1, 100.0, 12541, 970317, 17184, 12, 1432, 'monthly', 960511)]

añadir dato nacimiento cliente

se conecta tabla de account con disposition y de ahi de account a client

In [20]:
%%sql
SELECT loan_id, A4, A9, A10, A11, c.birth_number, "1".date loan_date, amount, duration, payments, frequency, a.date acc_date
FROM LOAN "1", ACCOUNT a, DISTRICT d, CLIENT c, DISPOSITION dp
WHERE "1".account_id = a.account_id
AND a.district_id = d.A1
AND dp.account_id = a.account_id
AND dp.account_id = c.client_id
AND status in ("A","B")
ORDER BY amount, a.date DESC
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(5657, 86513, 5, 50.5, 8288, '436028', 940705, 4980, 12, 415, 'monthly', 931105),
 (6229, 1204953, 1, 100.0, 12541, '630410', 941002, 11736, 24, 489, 'monthly', 930422),
 (5103, 67298, 7, 43.8, 8561, '820523', 971128, 12540, 12, 1045, 'monthly', 960315),
 (5103, 67298, 7, 43.8, 8561, '820523', 971128, 12540, 12, 1045, 'monthly', 960315),
 (5718, 107911, 7, 50.9, 8240, '590110', 970901, 14028, 12, 1169, 'monthly', 951028),
 (6064, 157042, 9, 33.9, 8743, '736015', 940804, 14628, 12, 1219, 'monthly', 930701),
 (5897, 122603, 8, 80.0, 8991, '270113', 960925, 15192, 24, 633, 'monthly', 941110),
 (5599, 128118, 8, 85.3, 9317, '300505', 950321, 15420, 12, 1285, 'monthly', 930523),
 (6297, 323870, 1, 100.0, 10673, '736020', 941025, 17508, 12, 1459, 'monthly', 930801),
 (5740, 1204953, 1, 100.0, 12541, '615206', 950402, 18564, 12, 1547, 'monthly', 941125)]

ir jugando para ver los tipos y poder procesar

In [21]:
%%sql
select distinct type from disposition

 * sqlite:///data_berka.db
Done.


[('OWNER',), ('DISPONENT',)]

filtro por tipo

In [22]:
%%sql
SELECT loan_id, A4, A9, A10, A11, c.birth_number, "1".date loan_date, amount, duration, payments, frequency, a.date acc_date
FROM LOAN "1", ACCOUNT a, DISTRICT d, CLIENT c, DISPOSITION dp
WHERE "1".account_id = a.account_id
AND a.district_id = d.A1
AND dp.account_id = a.account_id
AND dp.account_id = c.client_id
AND dp.type = "OWNER"
AND status in ("A","B")
ORDER BY amount, a.date DESC
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(5657, 86513, 5, 50.5, 8288, '436028', 940705, 4980, 12, 415, 'monthly', 931105),
 (6229, 1204953, 1, 100.0, 12541, '630410', 941002, 11736, 24, 489, 'monthly', 930422),
 (5103, 67298, 7, 43.8, 8561, '820523', 971128, 12540, 12, 1045, 'monthly', 960315),
 (5718, 107911, 7, 50.9, 8240, '590110', 970901, 14028, 12, 1169, 'monthly', 951028),
 (6064, 157042, 9, 33.9, 8743, '736015', 940804, 14628, 12, 1219, 'monthly', 930701),
 (5897, 122603, 8, 80.0, 8991, '270113', 960925, 15192, 24, 633, 'monthly', 941110),
 (5599, 128118, 8, 85.3, 9317, '300505', 950321, 15420, 12, 1285, 'monthly', 930523),
 (6297, 323870, 1, 100.0, 10673, '736020', 941025, 17508, 12, 1459, 'monthly', 930801),
 (5740, 1204953, 1, 100.0, 12541, '615206', 950402, 18564, 12, 1547, 'monthly', 941125),
 (5651, 75232, 5, 41.7, 8980, '610320', 971208, 19248, 12, 1604, 'monthly', 961012)]

extraer los 2 primeros digitos del birth number

In [23]:
%%sql
SELECT loan_id, A4, A9, A10, A11, c.birth_number, SUBSTR(c.birth_number,0,3), "1".date loan_date, amount, duration, payments, frequency, a.date acc_date
FROM LOAN "1", ACCOUNT a, DISTRICT d, CLIENT c, DISPOSITION dp
WHERE "1".account_id = a.account_id
AND a.district_id = d.A1
AND dp.account_id = a.account_id
AND dp.account_id = c.client_id
AND status in ("A","B")
ORDER BY amount, a.date DESC
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(5657, 86513, 5, 50.5, 8288, '436028', '43', 940705, 4980, 12, 415, 'monthly', 931105),
 (6229, 1204953, 1, 100.0, 12541, '630410', '63', 941002, 11736, 24, 489, 'monthly', 930422),
 (5103, 67298, 7, 43.8, 8561, '820523', '82', 971128, 12540, 12, 1045, 'monthly', 960315),
 (5103, 67298, 7, 43.8, 8561, '820523', '82', 971128, 12540, 12, 1045, 'monthly', 960315),
 (5718, 107911, 7, 50.9, 8240, '590110', '59', 970901, 14028, 12, 1169, 'monthly', 951028),
 (6064, 157042, 9, 33.9, 8743, '736015', '73', 940804, 14628, 12, 1219, 'monthly', 930701),
 (5897, 122603, 8, 80.0, 8991, '270113', '27', 960925, 15192, 24, 633, 'monthly', 941110),
 (5599, 128118, 8, 85.3, 9317, '300505', '30', 950321, 15420, 12, 1285, 'monthly', 930523),
 (6297, 323870, 1, 100.0, 10673, '736020', '73', 941025, 17508, 12, 1459, 'monthly', 930801),
 (5740, 1204953, 1, 100.0, 12541, '615206', '61', 950402, 18564, 12, 1547, 'monthly', 941125)]

In [24]:
%%sql
SELECT count(*)
FROM LOAN "1", ACCOUNT a
WHERE "1".account_id = a. account_id

 * sqlite:///data_berka.db
Done.


[(682,)]

consultar presatmos superiores a esa cantidad

In [25]:
%%sql
select count(*)
from loan
where amount >153957.2857142857

 * sqlite:///data_berka.db
Done.


[(277,)]

11. Domiciliaciones en cuentas de actualización semanal de alquiler en cuentas de Praga

In [26]:
%%sql
SELECT order_id, p.account_id, amount
FROM PERMANENT_ORDER p, ACCOUNT a, DISTRICT d
WHERE p.account_id = a.account_id
and a.district_id = d.A1
and k_symbol = "LEASING"
and frequency = "weekly"
and A3 = "Prague"

 * sqlite:///data_berka.db
Done.


[(29514, 72, 2923),
 (30250, 573, 2249),
 (30944, 1056, 2531.7),
 (34056, 3141, 1087),
 (35237, 3959, 1680.6),
 (40429, 7465, 1060.6)]


12. Domiciliaciones (*órdenes*) semanales para alquiler (*leasing*) en cuentas de Praga

 - Ordenes permanentes de alquiler (*leasing*) y semanales (*weekly*)
 - Relación con cuentas que tienen región Praga

In [27]:
%%sql
SELECT order_id, k_symbol, bank_to, a.account_id, a.frequency, d.A2, d.A3
FROM PERMANENT_ORDER p, ACCOUNT a, DISTRICT d
WHERE p.account_id = a.account_id and a.district_id = d.A1
    and k_symbol = "LEASING" and a.frequency = "weekly" and d.A3 = "Prague"

 * sqlite:///data_berka.db
Done.


[(29514, 'LEASING', 'OP', 72, 'weekly', 'Hl.m. Praha', 'Prague'),
 (30250, 'LEASING', 'CD', 573, 'weekly', 'Hl.m. Praha', 'Prague'),
 (30944, 'LEASING', 'IJ', 1056, 'weekly', 'Hl.m. Praha', 'Prague'),
 (34056, 'LEASING', 'OP', 3141, 'weekly', 'Hl.m. Praha', 'Prague'),
 (35237, 'LEASING', 'AB', 3959, 'weekly', 'Hl.m. Praha', 'Prague'),
 (40429, 'LEASING', 'ST', 7465, 'weekly', 'Hl.m. Praha', 'Prague')]

13. Transacciones relacionadas con pensiones (DUCHOD) en Plzen desde el banco MN

(limitar a 10)

In [28]:
%%sql
SELECT trans_id, a.account_id, a.frequency, t.type, k_symbol, bank
FROM TRANS t, ACCOUNT a, DISTRICT d
WHERE t.account_id = a.account_id and a.district_id = d.A1
  and k_symbol = "DUCHOD" and bank = "MN" and d.A2 like "Plzen%"  
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(47969, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47970, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47971, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47972, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47973, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47974, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47975, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47976, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (47977, 161, 'monthly', 'PRIJEM', 'DUCHOD', 'MN'),
 (1098524, 3753, 'monthly', 'PRIJEM', 'DUCHOD', 'MN')]

# Subconsultas

14. Prestamos cuya cantidad es superior a la media de los clientes de Praga

- Cantidad media prestada a los cuentas de Praga 

numero de prestamos

In [29]:
%%sql
SELECT count(*)
FROM LOAN "1", ACCOUNT a, DISTRICT d
WHERE "1".account_id = a.account_id
AND a.district_id = d.A1
AND d.A3 = "Prague"


 * sqlite:///data_berka.db
Done.


[(84,)]

media

In [30]:
%%sql
SELECT avg("1".amount)
FROM LOAN "1", ACCOUNT a, DISTRICT d
WHERE "1".account_id = a.account_id
AND a.district_id = d.A1
AND d.A3 = "Prague"

 * sqlite:///data_berka.db
Done.


[(153957.2857142857,)]

- Consultar los préstamos superiores a esa cantidad

In [31]:
%%sql
select count(*)
from loan
where amount >153957.2857142857


 * sqlite:///data_berka.db
Done.


[(277,)]

Haciendolo como subsconsulta

In [32]:
%%sql
select count(*)
from loan
where amount > (select avg("1".amount)
  FROM LOAN "1", ACCOUNT a, DISTRICT d
  WHERE "1".account_id = a.account_id
  AND a.district_id = d.A1
  AND d.A3 = "Prague")

 * sqlite:///data_berka.db
Done.


[(277,)]

# Repaso de Join/Group By

## Objetivo 2.1

15 . Clientes de Praga con préstamos según su estado

- ¿Cuántos clientes hay en Praga?
- ¿Cuántas cuentas tienen asociados clientes en Praga?
- ¿cuántas de esas cuentas tienen asociados un préstamo?
- Necesito saber cuántos hay por cada estado (*status*)


In [33]:
%%sql 
SELECT status, count(*)
FROM CLIENT c, DISTRICT d, DISPOSITION dp, LOAN l 
WHERE c.district_id = d.A1 and c.client_id = dp.client_id
and l.account_id = dp.account_id and d.A3 = "Prague"
GROUP BY status

 * sqlite:///data_berka.db
Done.


[('A', 37), ('B', 3), ('C', 55), ('D', 3)]

## Ejercicio

> cantidad de dinero medio en transacciones de clientes propietarios de cuentas en Kolin (owned accounts)

In [34]:
%%sql
SELECT c.client_id, avg (amount) as av
from client c, district d, disposition ds, account a, trans t
WHERE c.district_id = d.A1 and c.client_id = ds.client_id
    and ds.account_id = a.account_id and a.account_id = t.account_id
and d.A2 = "Kolin" and ds.type = "OWNER" 
GROUP BY c.client_id
ORDER BY av DESC
LIMIT 10



 * sqlite:///data_berka.db
Done.


[(5249, 15611.831578947369),
 (3378, 14803.997402597402),
 (1388, 14464.587096774194),
 (6803, 13825.125153374232),
 (8091, 13227.722338204592),
 (3298, 13074.464857881138),
 (1806, 13037.759375),
 (4331, 12644.448051948051),
 (4974, 12023.291743119267),
 (3080, 11784.50460251046)]

> sanction interest if negative balance Cesky Krumlov


> stands for loan payment in Praha from the bank EF

# Extra: Left Join

1. ¿Cuántas cuentas hay? 4500
2. ¿Cuántas cuentas hay con préstamos? (inner join) 682
3. Cuentas con los datos de los préstamos si los hay
4. los ordenamos por cantidad


In [35]:
%%sql 
SELECT a.account_id, a.date, l.amount, l.status, l.payments, l.duration
FROM ACCOUNT a INNER OUTER JOIN LOAN l ON a.account_id = l.account_id
ORDER BY amount DESC




 * sqlite:///data_berka.db
(sqlite3.OperationalError) unknown or unsupported join type: INNER OUTER
[SQL: SELECT a.account_id, a.date, l.amount, l.status, l.payments, l.duration
FROM ACCOUNT a INNER OUTER JOIN LOAN l ON a.account_id = l.account_id
ORDER BY amount DESC]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


# DataSets

---

Dataset sobre las transacciones con Vyskov, extrayendo los años de las diferentes fechas. 

In [36]:
%%sql dataset << SELECT substr(a.date,0,3) acc_year, substr(t.date, 0, 3) year,amount, balance, k_symbol
FROM ACCOUNT a, TRANS t, DISTRICT d
WHERE a.district_id = d.A1 and d.A2 = "Vyskov"
AND a.account_id = t.account_id


 * sqlite:///data_berka.db
Done.
Returning data to local variable dataset


In [37]:
dataset.DataFrame().to_csv("dataset.csv")